Webpack

Tvrqvoise opened the pull request:
A whole bunch of stuff here, but these are the highlights:
 - Upgrade the build to Webpack
 - Remove Gulp
 - Upgrade babel version, add latest and stage 0 presets.  Tested this out by using static property initializers in App.jsx
 - Unversion the scripts, styles and images folders in public.  Sources for these things were moved to the src/ directory, and webpack now spits out its output into the public/scripts directory
 - Charts and Site now go through webpack too.
 - Fixed CSS minifying and prefixing

A couple of notes:
 - I'm not really sure how to test actual games.  I've got it running locally and all the pages seem to work fine, but I don't know how to actually start a game with players.  If you're interested in this change, you might want to install this repo locally and check it out yourself.  Biggest concern is that there's a 404 somewhere.
 - I don't know how you deploy to production, but that might be impacted by these changes.  Specifically, `gulp build` is now `npm run compile`, and `gulp` is now `npm run watch`.
 - I found a few missing images while doing this.  Do you know what happened to them?
 - Is there a reason why the game SASS imports its own copy of semantic UI instead of using the CDN one?

Tvrqvoise commented on src/frontend-scripts/components/section-main/Defaultmid.jsx:
```diff
@@ -28,7 +29,7 @@ const Defaultmid = ({ version, readPatchNotes, quickDefault }) => (
 		<PatchAlert
 			isActive={version.lastSeen && version.current.number !== version.lastSeen}
 			onClick={readPatchNotes} />
-		<img src="images/lizard17.png" alt="Secret Hitler logo" width="400" height="400" />
```
I couldn't find lizard17.png, even on master.

cozuya commented:
Wow thanks.  I was planning on working on this next week but will take a look here.  I'd like to work off of an ejected create-react-app webpack config as thats more or less becoming the standard.

* to test a game with players I'd recommend making the accounts that are on the buttons on Defaultmid and then making 5 sessions using Chrome MultiLogin.  Instructions are in the readme.

* I had to change some parts of semantic's innards to get it to do what I want (I can't remember specifics) so I just "sassed" it.

* Lizard17 has yet to be made :)  I typically throw that out there so I don't forget because I'm prepping for the next release which is probably tonight and then photoshop the next lizard as a near last step.

Tvrqvoise commented on src/scss/semantic.scss:
```diff
@@ -4750,36 +4750,6 @@ input::selection {
-
-/* Flag Sprite Based On http://www.famfamfam.com/lab/icons/flags/ */
-
-/*******************************
```
The `.flags` code seemed to be unused, and webpack was puking when trying to resolve the image.  Seemed safe to remove.  I can put this back if needed.  Same change was made in semantic-full.

jbasrai commented:
These changes look great! The webpack bundle times are noticeably faster.

- For my benefit, can you explain what unversioning the files in the public folder means? Guessing it has to do with browser caching, but would love clarification and what webpack's role in this is.
- Could you explain the fix to CSS minifying/prefixing? What was the problem and what did you do resolve it? Again, this is just for my curiosity.
- It looks like Babel's [latest](https://babeljs.io/docs/plugins/preset-latest/) preset is deprecated in favor of the [env](https://babeljs.io/docs/plugins/preset-env/) preset. Should we consider using the newer preset?
- It looks like some images aren't being captured by webpack. Specifically the ones referenced by the static pages, such as on the homepage and the how-to-play page. Do we need to add an entry point for `src/images`?
- Do you know why some images, like `/src/images/fascist-logo.png`, get included in the bundle as their own file, while most other images are converted inline to the stylesheet? Is it related to the size of the image or something?
- `/public/scripts/game.js` seems to take a while to run and the absence of stylesheets while it loads makes for an abrupt transition. To illustrate, when I open the `/observe` page, for a second I see [rendered html without styling](https://user-images.githubusercontent.com/5355316/27244673-73b04514-529c-11e7-9fdc-d2555461fe9f.png) before the page is loaded. I'm not sure what an appropriate solution is, do you think we should have some boot-up CSS that loads before the script is run?

I started a dev game and everything looks fine.

On a side note, since we're seeing an influx of forks/contributors I'd like to expose a small script for first-time run that populates the database and makes it easier to get test games started.

jbasrai commented on .babelrc:
```diff
@@ -1,6 +1,7 @@
\ No newline at end of file
+  	"latest",
+  	"react",
+    "babel-preset-stage-0"
```
Minor point, and this existed before these changes, but there are some mixed spaces and tabs here.

cozuya replied:
Weird wonder how that happened.  I've addressed now.

cozuya commented:
> On a side note, since we're seeing an influx of forks/contributors I'd like to expose a small script for first-time run that populates the database and makes it easier to get test games started.

That would be great, I haven't spent as much time in mongo as I should.  I'm also going to add a contributers.txt to /docs.  And uh stop doing commits right on master, after this release I'll do extra stuff in feature branches.

Tvrqvoise commented:
> For my benefit, can you explain what unversioning the files in the public folder means? Guessing it has to do with browser caching, but would love clarification and what webpack's role in this is.

Webpack will follow the image URLs in the rest of the bundle and optimize them, so it needs to control the output in order for the images to be properly compressed, fingerprinted, etc.

> Could you explain the fix to CSS minifying/prefixing? What was the problem and what did you do resolve it? Again, this is just for my curiosity.

I changed the CSS minification to css-nano, and added autoprefixer.  I'm not seeing #79 or #63, so I'm speculatively hoping that my changes resolved them, though I did not target them specifically.

> It looks like Babel's latest preset is deprecated in favor of the env preset. Should we consider using the newer preset?

Good point.  `latest` supersedes `es2015`, which is why I installed it, but you're right that env would be a better choice.  I'll play around with it a little.

> It looks like some images aren't being captured by webpack. Specifically the ones referenced by the static pages, such as the home page and the how-to-play page. Do we need to add an entry point for src/images?

We could, but then the images wouldn't be properly compressed, as currently Webpack doesn't have any knowledge of what's going on in Jade.  I'll see if I can't just reference them from CSS for now.  Other solutions I see here would be:
 - Use something like https://github.com/pugjs/pug-loader
 - Do the other pages in React
 - Precompile the other pages at build time

> Do you know why some images, like /src/images/fascist-logo.png, get included in the bundle as their own file, while most other images are converted inline to the stylesheet? Is it related to the size of the image or something?

This happens via Webpack's `url-loader`, which takes a binary file like an image and turns it into a base-encoded URL.  The relevant bit is here:

```js
{
  test: /\.(jpe?g|png|gif|svg)/,
  loaders: [
    {
      loader: 'url-loader',
      query: {
        hash: 'sha512',
        digest: 'hex',
        name: '[name]-[hash].[ext]',
        limit: 32000
      }
    }
  ]
}
```

Note the `limit: 32000` setting.  This is our way of telling the `url-loader`, "only bundle images if they're below 32k in size".  Everything bigger than that gets loaded externally via Webpack's `file-loader`.

> /public/scripts/game.js seems to take a while to run and the absence of stylesheets while it loads makes for an abrupt transition. To illustrate, when I open the /observe page, for a second I see rendered html without styling before the page is loaded. I'm not sure what an appropriate solution is, do you think we should have some boot-up CSS that loads before the script is run?

I'll play around with this and see what I can do.

Tvrqvoise commented:
Regarding babel-preset-env, what browsers would we want to support?

Tvrqvoise commented:
So, the issue with the images in pug seems to be due to the changelog, which I'm discovering has a lot of similarities with `Changelog.jsx`.  Would anyone object to me merging the two templates?

cozuya commented:
Could you pretty easily merge the 2 pages by putting the text and headers into a variable (version.js would work for this) and having react and pug read from them separately.  Pug has variable support of course.

Tvrqvoise commented:
There's a limited amount of formatting in the release notes (including embedded images, which was what that @jbasrai's comment was about).  Either way, I think it's somewhat beyond the scope of this change -- as it stands, it works.  It might be better to back up and take full consideration of what you want out of the changelog, the best way for users to get that information, and then let the solution follow that.

cozuya commented:
So unfortunately I don't think I can bring this in.  Its mostly on me - I did not have collab requirements in docs until after you started work on this.  Reason being is I assigned the issue to me and was planning on working on it myself.  This isn't really what I was thinking - what I was thinking was literally clone and eject the current version of CRA and then integrate this project in there as opposed to trying to integrate webpack into this project and its structure.  I think that would be the most future forward way of doing it.  Also this PR doesn't seem to work with Windows ><.

If you would like try that, please do!  If not, I will get started on that pretty soon.  Let me know what you think.

Tvrqvoise commented:
Go for it, it will be cool to see what you come up with.
